File: challenge-01-order-book-rust/solution/heap_impl.rs
use crate::types::OrderBook;
use std::collections::{BinaryHeap, HashSet};

pub struct HeapOb{
    id_st: HashSet<u64>,
    bids: BinaryHeap<Level>,
    asks: BinaryHeap<Level>
}

#[derive(PartialEq, Eq)]
struct Level{
    price: i64,
    order_id: u64
}

impl PartialOrd for Level {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        self.price.partial_cmp(&other.price)
    }
}

impl Ord for Level {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.price.cmp(&other.price)
    }
}

impl OrderBook for HeapOb{
    fn new() -> Self{
        Self{
            id_st: HashSet::new(),
            bids: BinaryHeap::new(),
            asks: BinaryHeap::new(),
        }
    }

    fn add_order(&mut self, id: u64, side: i32, price: i64, quantity: i64){
        self.id_st.insert(id);
        if side == 0{
            self.bids.push(Level {price: price, order_id: id});
        }
        else {
            self.asks.push(Level {price: - price, order_id: id});
        }
    }

    fn cancel_order(&mut self, id: u64){
        self.id_st.remove(&id);

        while !self.bids.is_empty(){ // ensure that the best bid is still valid
            if let Some(level) = self.bids.peek(){
                if self.id_st.contains(&level.order_id){
                    break;
                }
            }
            self.bids.pop();
        }

        while !self.asks.is_empty(){ // ensure that the best ask is still valid
            if let Some(level) = self.asks.peek(){
                if self.id_st.contains(&level.order_id){
                    break;
                }
            }
            self.asks.pop();
        }
    }

    fn best_bid(&self) -> i64{
        match self.bids.peek(){
            Some(level) => level.price,
            None => 0
        }
    }

    fn best_ask(&self) -> i64{
        match self.asks.peek(){
            Some(level) => - level.price,
            None => 0
        }
    }
}


#[cfg(test)]
mod heap_ut{
    use super::*;
    use crate::ut::{self};

    #[test]
    fn test_bid(){
        ut::ut_common::test_bid(HeapOb::new());
    }

    #[test]
    fn test_ask(){
        ut::ut_common::test_ask(HeapOb::new());
    }

    #[test]
    fn test_cancel(){
        ut::ut_common::test_cancel(HeapOb::new());
    }

    #[test]
    fn test_empty(){
        ut::ut_common::test_empty(HeapOb::new());
    }

    #[test]
    fn test_duplicate(){
        ut::ut_common::test_duplicate_price(HeapOb::new());
    }
}

File: challenge-01-order-book/solution/heap_impl.cpp
#include "solution.h"
#include <unordered_set>
#include <queue>

namespace hftu {
    struct Order {
        int64_t price;
        uint64_t id;
    };

    struct Ascending {
        bool operator()(const Order &a, const Order &b) { return a.price > b.price; };
    };

    struct Descending {
        bool operator()(const Order &a, const Order &b) { return a.price < b.price; };
    };

    class Impl{
        public:
            Impl(){

            }

            inline void add_order(uint64_t id, int side, int64_t price, int64_t quantity){
                orders_.insert(id);
                if (side){ //asks
                    asks_.push({price, id});
                }
                else {
                    bids_.push({price, id});
                }
            }

            inline void cancel_order(uint64_t id){
                orders_.erase(id);

                while (!asks_.empty() && orders_.find(asks_.top().id) == orders_.end()){
                    asks_.pop();
                }

                while (!bids_.empty() && orders_.find(bids_.top().id) == orders_.end()){
                    bids_.pop();
                }
            }

            inline int64_t best_bid() const {
                if (bids_.empty()) return 0;
                return bids_.top().price;
            }

            inline int64_t best_ask() const{
                if (asks_.empty()) return 0;
                return asks_.top().price;
            }

        private: 
            std::unordered_set<uint64_t> orders_;
            std::priority_queue<Order, std::vector<Order>, Descending> bids_;
            std::priority_queue<Order, std::vector<Order>, Ascending> asks_;            
    };


    void OrderBook::add_order(uint64_t id, int side, int64_t price, int64_t quantity) {
        this->impl->add_order(id, side, price, quantity);
    }

    void OrderBook::cancel_order(uint64_t id) {
        this->impl->cancel_order(id);
    }

    int64_t OrderBook::best_bid() const {
        return this->impl->best_bid();
    }

    int64_t OrderBook::best_ask() const {
        return this->impl->best_ask();
    }   

    OrderBook::OrderBook(){
        this->impl = new Impl();
    }

    OrderBook::~OrderBook(){
        delete this->impl;   
    }
} 
